tensorflow/contrib/data/python/ops/get_single_element.py

```diff
@@ -91,8 +91,11 @@ def reduce_dataset(dataset, reducer):
     raise TypeError("`dataset` must be a `tf.data.Dataset` object.")
 
   # The sentinel dataset is used in case the reduced dataset is empty.
+  initial_state = reducer.init_func(np.int64(0))
+  if not isinstance(initial_state, tuple):
+      initial_state = (initial_state,)
   sentinel_dataset = dataset_ops.Dataset.from_tensors(
-      reducer.finalize_func(reducer.init_func(np.int64(0))))
+      reducer.finalize_func(*initial_state))
```

Comment:
Good point, I've looked into it and if the output of `reducer.init_func()` is a dictionary, `initial_state` will not be a tuple so we will convert it to a tuple with one element which will be passed to:
```python
reducer.finalize_func(*initial_state)
```
so it should be ok.


I've looked into the code you linked:
```python
if not _should_unpack_args(nested_args): 
  nested_args = (nested_args,) 
```

and the actual code of `_should_unpack_args` is:
```python
def _should_unpack_args(args):
  """Returns `True` if `args` should be `*args` when passed to a callable."""
  return type(args) is tuple  # pylint: disable=unidiomatic-typecheck
```

so this is the same as what I did.

Let me know what you think, I can definitely add a test with a dictionary state, and also refactor with a function named `_should_unpack_args` for clarity.

Reply:
The only difference would be in using `type(...) is tuple` or `isinstance(..., tuple)`.

I guess using `type` makes more sense here since users could pass an object inheriting from tuple that shouldn't be expanded.

Reply:
@mrry: let me know what you think :)